Date strings are incorrectly returned as datetime objects after being stored and retrieved from Redis cache

### Bug Description

A date string such as `2001-11-08` is converted to a datetime object after being stored and retrieved from the Redis cache.

#### Steps to Reproduce

This is easily reproduced by this test
```
def test_datetime_string_stays_string(self):
        date = {"birthday": "2001-11-08"}
        decoded = FidesopsRedis.decode_obj(FidesopsRedis.encode_obj(date))
        assert decoded == date
```
```
AssertionError: assert {'birthday': datetime.datetime(2001, 11, 8, 0, 0)} == {'birthday': '2001-11-08'}
```

#### Root cause
The `_custom_decoder` in `cache.py` converts any date string to a datetime object after it is read from the cache. This change was introduced as part of https://github.com/ethyca/fides/pull/2643
```
def _custom_decoder(json_dict: Dict[str, Any]) -> Dict[str, Any]:
    for k, v in json_dict.items():
        try:
            json_dict[k] = datetime.fromisoformat(v)
            continue
        except (TypeError, ValueError):
            pass
```

### Expected behavior

The object should be stored and retrieved without converting the date string to a datetime object.

#### Potential fix

Store any date string with a prefix like we do when storing `bytes` or `ObjectId` so that they are not converted to datetime objects after being read from the cache.

### Additional context

This issue has the potential to affect SaaS erasure requests since we can't build a valid payload to update a collection after the access request data has been stored and retrieved from the cache.
```TypeError: Object of type datetime is not JSON serializable```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -40,6 +40,7 @@ The types of changes are:
 
 * Fix support for "redis.user" setting when authenticating to the Redis cache [#2666](https://github.com/ethyca/fides/pull/2666)
 * Fix error with the classify dataset feature flag not writing the dataset to the server [#2675](https://github.com/ethyca/fides/pull/2675)
+* Allow string dates to stay strings in cache decoding [#2695](https://github.com/ethyca/fides/pull/2695)
 * Admin UI
   * Remove Identifiability (Data Qualifier) from taxonomy editor [2684](https://github.com/ethyca/fides/pull/2684)
 

--- a/src/fides/api/ops/util/cache.py
+++ b/src/fides/api/ops/util/cache.py
@@ -21,6 +21,7 @@ RedisValue = Union[bytes, float, int, str]
 _connection = None
 
 ENCODED_BYTES_PREFIX = "quote_encoded_"
+ENCODED_DATE_PREFIX = "date_encoded_"
 ENCODED_MONGO_OBJECT_ID_PREFIX = "encoded_object_id_"
 
 
@@ -31,7 +32,7 @@ class CustomJSONEncoder(json.JSONEncoder):
         if isinstance(o, bytes):
             return f"{ENCODED_BYTES_PREFIX}{quote(o)}"
         if isinstance(o, (datetime, date)):
-            return o.isoformat()
+            return f"{ENCODED_DATE_PREFIX}{o.isoformat()}"
         if isinstance(o, ObjectId):
             return f"{ENCODED_MONGO_OBJECT_ID_PREFIX}{str(o)}"
         if isinstance(o, object):
@@ -47,17 +48,13 @@ class CustomJSONEncoder(json.JSONEncoder):
 
 def _custom_decoder(json_dict: Dict[str, Any]) -> Dict[str, Any]:
     for k, v in json_dict.items():
-        try:
-            json_dict[k] = datetime.fromisoformat(v)
-            continue
-        except (TypeError, ValueError):
-            pass
-
         if isinstance(v, str):
             # The mongodb objectids couldn't be directly json encoded so they are converted
             # to strings and prefixed with encoded_object_id in order to find during decodeint.
             if v.startswith(ENCODED_MONGO_OBJECT_ID_PREFIX):
                 json_dict[k] = ObjectId(v[18:])
+            if v.startswith(ENCODED_DATE_PREFIX):
+                json_dict[k] = datetime.fromisoformat(v[13:])
             # The bytes from secrets couldn't be directly json encoded so it is url
             # encode and prefixed with quite_encoded in order to find during decodeint.
             elif v.startswith(ENCODED_BYTES_PREFIX):

--- a/tests/ops/util/test_cache.py
+++ b/tests/ops/util/test_cache.py
@@ -10,6 +10,7 @@ from bson.objectid import ObjectId
 
 from fides.api.ops.util.cache import (
     ENCODED_BYTES_PREFIX,
+    ENCODED_DATE_PREFIX,
     ENCODED_MONGO_OBJECT_ID_PREFIX,
     FidesopsRedis,
 )
@@ -121,7 +122,15 @@ class TestCustomJSONEncoder:
             (b"some value", f'"{ENCODED_BYTES_PREFIX}some%20value"'),
             (
                 datetime(2023, 2, 14, 20, 58),
-                f'"{datetime(2023, 2, 14, 20, 58).isoformat()}"',
+                f'"{ENCODED_DATE_PREFIX}{datetime(2023, 2, 14, 20, 58).isoformat()}"',
+            ),
+            (
+                {"a": datetime(2023, 2, 14, 20, 58)},
+                f'{{"a": "{ENCODED_DATE_PREFIX}{datetime(2023, 2, 14, 20, 58).isoformat()}"}}',
+            ),
+            (
+                {"a": {"b": datetime(2023, 2, 14, 20, 58)}},
+                f'{{"a": {{"b": "{ENCODED_DATE_PREFIX}{datetime(2023, 2, 14, 20, 58).isoformat()}"}}}}',
             ),
             ({"a": "b"}, '{"a": "b"}'),
             ({"a": {"b": "c"}}, '{"a": {"b": "c"}}'),
@@ -152,16 +161,28 @@ class TestCustomDecoder:
         [
             (f'{{"a": "{ENCODED_BYTES_PREFIX}some%20value"}}', {"a": b"some value"}),
             (
-                f'{{"a": "{datetime(2023, 2, 17, 14, 5).isoformat()}"}}',
-                {"a": datetime(2023, 2, 17, 14, 5)},
-            ),
-            (
                 f'{{"a": "{ENCODED_MONGO_OBJECT_ID_PREFIX}507f191e810c19729de860ea"}}',
                 {"a": ObjectId("507f191e810c19729de860ea")},
             ),
+            (
+                f'{{"a": "{ENCODED_DATE_PREFIX}{datetime(2023, 2, 17, 14, 5).isoformat()}"}}',
+                {"a": datetime(2023, 2, 17, 14, 5)},
+            ),
+            (
+                f'{{"a": {{"b": {{"c": "{ENCODED_DATE_PREFIX}{datetime(2023, 2, 17, 14, 5).isoformat()}"}}}}}}',
+                {"a": {"b": {"c": datetime(2023, 2, 17, 14, 5)}}},
+            ),
+            (
+                '{"birthday": "2001-11-08"}',
+                {"birthday": "2001-11-08"},
+            ),
+            (
+                '{"a": {"b": {"birthday": "2001-11-08"}}}',
+                {"a": {"b": {"birthday": "2001-11-08"}}},
+            ),
         ],
     )
-    def test_decode_bytes(self, value, expected):
+    def test_cache_decode(self, value, expected):
         cache = FidesopsRedis()
         assert cache.decode_obj(value) == expected
 

